locust/test/test_main.py

```diff
@@ -647,21 +887,34 @@ def test_autostart_w_run_time(self):
                 stderr=PIPE,
                 text=True,
             )
-            gevent.sleep(2.8)
+
             try:
+                poll_until(lambda: web_interface_ready("localhost", port))
+
                 response = requests.get(f"http://localhost:{port}/")
-            except Exception:
-                pass
-            _, stderr = proc.communicate(timeout=4)
-            self.assertIn("Starting Locust", stderr)
-            self.assertIn("Run time limit set to 3 seconds", stderr)
-            self.assertIn("Shutting down ", stderr)
-            self.assertNotIn("Traceback", stderr)
-            # check response afterwards, because it really isn't as informative as stderr
-            d = pq(response.content.decode("utf-8"))
+                self.assertEqual(200, response.status_code)
 
-            self.assertEqual(200, response.status_code)
-            self.assertIn('"state": "running"', str(d))
+                def process_finished():
+                    return proc.poll() is not None
+
+                poll_until(process_finished, timeout=10)
```

Comment:
No need to poll for process finishing, that's what proc.communicate does anyway.